### PR TITLE
Save workflow

### PR DESCRIPTION
**Pull request by catscanner**
Moves the service name out of parameters and into the step config.
Saves workflow and runids into versions when the deployment has been accepted by the slingshot engine.

### CODE REVIEW

File: slingshotapi/backend.py [hunk 1]
```diff
@@ -344,7 +344,7 @@ def save(version_region, version_data, deployment_id, plugin_image, service_id):
             deployment_id=deployment_id,
             version_plugin_image=plugin_image,
             version_region=version_region,
-            version_payload=json.dumps(version_data['payload'])
+            version_payload=json.dumps(version_data)
         )
```
Comment (gnhenderson) [inline slingshotapi/backend.py]: This level of nesting seems unnecessary - `'version_payload': {...}` is implicitly the version payload whilst
```
'version_payload': {
    'payload': {...}
}
```
will add more complexity to any later data retrieval attempts.

File: slingshotapi/deploy.py [hunk 2]
```diff
@@ -61,24 +78,29 @@ def deploy(cls, deployment_data):
...
     @classmethod
     def _process_deploys(cls, deployment):
+        """
+        Produce region-specific deployment data payloads updated with region-specific overrides
+
+        :param deployment: json contents of a slingshot yaml
+        :type deployment: dict
+
+        :param overrides: The contents of a slingshot yaml file's override section.
+        :type overrides: dict
```
Comment (gnhenderson) [inline slingshotapi/deploy.py]: param `overrides` is not passed to the method and does not need to be in the doc string.

File: slingshotapi/backend.py (hunk 1 above)
Comment (catscanner) [inline slingshotapi/backend.py]: There is no version_data['payload'] at this point. As the json processing has been done a different way, it's structured slightly differently.

Comment (gnhenderson) [inline slingshotapi/backend.py]: `version_payload` should be the contents of the `parameters:` dict for a single service. At the moment, it is the entire `pipeline` which is being saved as a version. Which means if there are 2 or more services in the pipeline they are being saved under the payload for a single service which is not how the data model is setup. The deployments table contains the payload sent to the API, which includes the entire pipeline. The versions table is only looking to snapshot the payload (`parameters` in this case) which apply to a single service.

The data being sent to the `save(.., version_data, ...)` method in `pipeline[endpoint]` is too broad and needs to be narrowed so it is only the `parameters` for an individual service which are being sent.

File: tests/test_backend.py [hunk 3]
```diff
@@ -191,9 +191,11 @@ def test_return_all_project_service_versions():
             "version_deployed_on": None,
             "version_id": 1,
             "version_payload": {
-                "loadbalancer": False,
-                "service_definition": {},
-                "task_definition": {}
+                "payload": {
```
Comment (gnhenderson) [inline tests/test_backend.py]: The shape of the `version_data` input data no longer represents what the application is passing to the database and `"version_payload": {"payload": {...}}` does not represent the shape of the expected output, e.g. the key `"payload"` does not exist in the returned json of the `ManageProject.service_versions()` method.

Due to the refactoring of the method which saves `version_payload` data there may be a question mark over what the expected inputs and outputs for these tests now looks like.

File: slingshotapi/deploy.py [hunk 4]
```diff
@@ -30,28 +31,43 @@ class DatabaseConnectionError(Exception):
     def deploy(cls, deployment_data):
         default_region = flask.current_app.config.get('SLINGSHOT_ENGINE', {}).get('default_endpoint', None)
         regions = deployment_data.get('regions', default_region) or default_region
```
Comment (gnhenderson) [inline slingshotapi/deploy.py]: `default_region` and `regions` are not used in the `deploy()` method and can be removed (as all region processing is handled by the `_process_deploys()` method).

File: tests/test_backend.py [hunk 5]
```diff
@@ -296,9 +298,11 @@ def test_return_all_service_versions_for_a_deployment(create_db, deployment_data
                 "version_deployed_on": None,
                 "version_id": 1,
                 "version_payload": {
-                    "loadbalancer": False,
-                    "service_definition": {},
-                    "task_definition": {}
+                    "payload": {
+                        "loadbalancer": False,
```
Comment (gnhenderson) [inline tests/test_backend.py]: ditto

File: tests/test_backend.py [hunk 6]
```diff
@@ -336,9 +340,11 @@ def test_return_a_service_version():
         "version_deployed_on": None,
         "version_id": version_id,
         "version_payload": {
-            "loadbalancer": False,
-            "service_definition": {},
-            "task_definition": {}
+            'payload': {
+                "loadbalancer": False,
```
Comment (gnhenderson) [inline tests/test_backend.py]: ditto

**Comment by gnhenderson** [comment]
Looking okay other than the data for the version tests: https://github.com/Skyscanner/slingshot-api/pull/30/files#diff-59ebb30a22811df4851a47d756309ba8

**Comment by gnhenderson** [comment]
LGTM

**Pull request merged**